[TEST] Filter Dropdown 테스트 코드 작성

EarlyHail opened the pull request:
- Filter Dropdown 버튼 테스트 코드 작성
  - Filter를 렌더링 후, Filters를 찾아 클릭 후, Open Issues를 확인

## 질문
```
describe('Something', () => {
  render(<Filter></Filter>);
  test('Initialize', async () => {
    //check initialization......
  });
  test('Filter Dropdown Btn Click', async () => {
    //check click event with trigger......
  });
});
```
이런식으로 같은 describe 내의 test suits에 사용되는 render를 공유하려고 했는데 
```
describe('Something', () => {
  test('Initialize', async () => {
    render(<Filter></Filter>);
    //check initialization......
  });
  test('Filter Dropdown Btn Click', async () => {
    render(<Filter></Filter>);
    //check click event with trigger......
  });
});
```
이렇게 따로 해야되더라구요.... 당연히 위에 방법도 될줄 알았는데 혹시 안되는 이유가 있을까요? 

## Issues
### .png파일을 jest가 로딩하지 못하여 에러 발생
   ![image](https://user-images.githubusercontent.com/55068119/97784140-b26bd380-1bdf-11eb-9a11-b8b9ff1a655f.png)
   Filter안의 FilterInputBox에서 png파일을 import하는데 import 문에서 에러가 발생합니다.
   계속 해결책을 찾아보다가 해결이 안되서 이미지를 제거하고 테스트를 짰는데 

   @lacomaco 영근님이 Container랑 Presentor랑 나누자고 하신 이유에 이런 것도 포함되어 있는건가요....???


### console.error 발생
   ![image](https://user-images.githubusercontent.com/55068119/97784204-2312f000-1be0-11eb-8f93-6d8bd1fc571a.png)

   이건 테스트 suits pass와 연관이 없는 에러인것 같긴 한데...
   Console로 나오는걸 보니 출력문인데 axiosMaker.test.js에 콘솔문이 없네요
   에러 내용을 보니 axiosMaker.js 또는 axiosMaker.test.js의 `windows` 관련 에러인것 같습니다.
   처음엔 안떴던것 같은데 png 이슈를 수정하느라 이것저것 해본것 때문에 생긴건 아닌지....
   확인 부탁드립니다😂😂😂

lacomaco committed:
Merge pull request #86 from boostcamp-2020/dev

[Week1] 최종배포

mon823 committed:
Update README.md

mon823 committed:
Update README.md

EarlyHail committed:
[TEST] Filter Dropdown 버튼 테스트

- Filter Dropdown 버튼 테스트 코드 작성
  - Filter를 렌더링 후, Filters를 찾아 클릭 후, Open Issues를 확인

### Issues
- console.error 발생
- .png파일을 jest가 로딩하지 못하여 에러 발생

EarlyHail committed:
[TEST] Filter Dropdown 버튼 테스트케이스 추가

- Filter Dropdown 버튼 테스트케이스 추가
   클릭으로 dropdown render 후 버튼 클릭 시 or dropdown 외부 클릭 시 close 되는지 test

lacomaco commented on frontEnd/src/__tests__/FilterDropdown.test.js:
```diff
@@ -0,0 +1,47 @@
+
+describe('Filter Dropdown Test', () => {
+  test('Initialize', async () => {
+    render(<Filter></Filter>);
```
https://testing-library.com/docs/react-testing-library/api#cleanup
Jest와 같이 global afterEach를 지원하는 testing framework 에서는 테스트마다 자동으로 cleanUp메소드가 호출된다고 합니다.

중복되는 렌더링 부분은
```js
beforeEach(()=>{
  render(<Filter></Filter>);
});
```
구문을 사용하면 중복을 줄일 수 있을것 같아요

lacomaco reviewed (commented):
.png를 로드하지 못하는 에러는 이 부분을 통해서 해결할 수 있을것 같아요
 
https://github.com/facebook/jest/issues/2663#issuecomment-341384494

axiosMaker에서 에러가 나는 부분은 jsdom에서 navigation을 구현하지 않아 생긴 부분인데 테스트 통과를 위해서 실제 코드를 지우는건 조금 아닌것 같아 우선은 방치해뒀습니다. 모킹을 통해서 에러 메시지가 발생하지 않게 처리할게욤 

.png 로드할때 생기는 에러 부분이랑 axiosMaker에서 불필요한 console.error 메시지가 발생하는 부분은 제가 고쳐서 PR 날릴게여

lacomaco commented on frontEnd/src/__tests__/FilterDropdown.test.js:
```diff
@@ -0,0 +1,47 @@
+    fireEvent.click(screen.getByText(/Filters/g));
+    await waitFor(() => {
+      try {
+        screen.getByText('Open Issues');
```
이 부분은 Open Issues가 없는것을 기대하고 tryCatch 구문에 담아두신 거죵??
getByText는 해당하는 텍스트가 없으면 바로 error을 던지지만 
queryBy구문은 해당하는 텍스트가 없으면 null을 반환해요
https://testing-library.com/docs/dom-testing-library/api-queries#queryby

이 부분은
```js
expect(screen.queryByText('Open Issues')).toBeNull();
```
으로 줄일 수 있을것 같아요.

lacomaco commented on frontEnd/src/__tests__/FilterDropdown.test.js:
```diff
@@ -0,0 +1,47 @@
+      }
+    });
+  });
+  test('Filter Dropdown Btn Click - Close with Click outside', async () => {
```
버튼 바깥쪽을 누르는 상황도 테스트 케이스에 넣으셨네요 엄청 꼼꼼하시네요 ㅋㅋ

EarlyHail committed:
[TEST] [REVIEW] FilterDropdown test 코드리뷰 반영

- 각 테스트마다 호출했던 render 함수를 beforeEach를 사용하여 한번의 호출로 대체
   --> render 중복 코드 삭제
- screen.getByText()와 try-catch를 이용한 테스트 함수를
   screen.queryByText()와 null Check로 대체

yyjjjj reviewed (approved):
주말 목표를 멋지게 달성하셨군요! 👍 👍
